packages/foundry/contracts/hooks/NftCheckHook.sol

```diff
@@ -0,0 +1,351 @@
+        address,
+        address pool,
+        TokenConfig[] memory _tokenConfigs,
+        LiquidityManagement calldata liquidityManagement
+    ) public override onlyVault returns (bool) {
+        if (liquidityManagement.enableDonation == false) {
+            revert PoolDoesNotSupportDonation();
+        }
+        tokenConfigs = _tokenConfigs;
+        poolAddress = pool;
+
+        emit NftCheckHookRegistered(address(this), pool);
+        return true;
+    }
+
+    function onBeforeInitialize(uint256[] memory /*exactAmountsIn*/, bytes memory /*userData*/) public override returns (bool) {
+        // Check if the hook owns the required NFT
+        if (IERC721(nftContract).ownerOf(nftId) != address(this)) {
+            revert DoesNotOwnRequiredNFT(address(this), nftContract, nftId);
+        }
+
+        // Check if the linked token is one of the pool tokens
+        bool linkedTokenFound = false;
+        for (uint256 i = 0; i < tokenConfigs.length; i++) {
+            if (address(tokenConfigs[i].token) == linkedToken) {
+                linkedTokenFound = true;
+                break;
+            }
+        }
+        if (!linkedTokenFound) {
+            revert LinkedTokenNotInPool(linkedToken);
+        }
+        // Record the initial liquidity amounts
+        recordInitialLiquidity(tokenConfigs[0].token.balanceOf(poolAddress), tokenConfigs[1].token.balanceOf(poolAddress));
+
+        return true;
+    }
+
+    /// @inheritdoc IHooks
+    function getHookFlags() public pure override returns (HookFlags memory) {
+        HookFlags memory hookFlags;
+        hookFlags.shouldCallAfterRemoveLiquidity = true;
+        hookFlags.shouldCallBeforeInitialize = true;
+        hookFlags.shouldCallBeforeSwap = true;
+        hookFlags.shouldCallBeforeAddLiquidity = true;
+        return hookFlags;
+    }
+
+    /// @inheritdoc IHooks
+    function onAfterRemoveLiquidity(
+        address,
+        address pool,
+        RemoveLiquidityKind kind,
+        uint256,
+        uint256[] memory,
+        uint256[] memory amountsOutRaw,
+        uint256[] memory,
+        bytes memory
+    ) public override onlyVault returns (bool, uint256[] memory hookAdjustedAmountsOutRaw) {
+        
+        // Ensure the first depositor has the same amount of both tokens after removal (no rug pulling allowed)
+        if (msg.sender == owner()) {
+            IERC20[] memory poolTokens = _vault.getPoolTokens(pool);
+
+            uint256 currentToken1Amount = poolTokens[0].balanceOf(msg.sender);
+            uint256 currentToken2Amount = poolTokens[1].balanceOf(msg.sender);
+
+            if (currentToken1Amount < initialToken1Amount) {
+                revert InsufficientLiquidityToRemove(msg.sender, currentToken1Amount, initialToken1Amount);
+            }
+            if (currentToken2Amount < initialToken2Amount) {
+                revert InsufficientLiquidityToRemove(msg.sender, currentToken2Amount, initialToken2Amount);
+            }
+        }
```

Comment:
This does not prevent a rug. In the instance that msg.sender was the user (which it's not, it's the vault). Consider the scenario:
1. NFT owner transfers NFT and then adds initial liquidity
2. NFT owner transfers BPT to a new wallet, that is now the `owner`.
3. New wallet calls remove liquidity, and bypasses the checks above.